File: java-demo/src/main/java/demo2/GraphEx.java
package demo2;

import akka.Done;
import akka.actor.ActorSystem;
import akka.actor.SystemGuardian;
import akka.stream.*;
import akka.stream.javadsl.*;

import java.util.concurrent.CompletionStage;
import java.util.stream.IntStream;

public class GraphEx {
    public static void main(String[] args) throws Exception {
        ActorSystem system = ActorSystem.create("demo2");
        ActorMaterializer materializer = ActorMaterializer.create(system);

        // create a stream for the following graph
        //                      f2
        // in -> f1 -> bcast /     \ merge -> f3 -> out
        //                   \  f4 /
        Sink<Object, CompletionStage<Done>> aSink = Sink.foreach(x -> System.out.println(x));
        CompletionStage<Done> run =
            RunnableGraph.fromGraph(GraphDSL.create(aSink, (builder, out) -> {
                IntStream stream = IntStream.range(1, 4);
                Outlet<Integer> in = builder.add(Source.from(() -> stream.iterator())).out();
                FlowShape<Integer, Integer> f1 =
                    builder.add(Flow.<Integer>create().map(x -> x + 10));
                FlowShape<Integer, Integer> f2 =
                    builder.add(Flow.<Integer>create().map(x -> x + 10));
                FlowShape<Integer, Integer> f3 =
                    builder.add(Flow.<Integer>create().map(x -> x + 10));
                FlowShape<Integer, Integer> f4 =
                    builder.add(Flow.<Integer>create().map(x -> x + 10));
                UniformFanOutShape<Integer, Integer> bcast = builder.add(Broadcast.create(2));
                UniformFanInShape<Integer, Integer> merge = builder.add(Merge.create(2));

                builder.from(in).via(f1).viaFanOut(bcast).via(f2).viaFanIn(merge).via(f3).to(out);
                builder.from(bcast).via(f4).toFanIn(merge);
                return ClosedShape.getInstance();
            })).run(materializer);

        run.whenComplete((done, throwable) -> {
            if (throwable != null) {
                System.out.println("Error occurred: " + throwable);
                system.terminate();
            }
            System.out.println("Done");
            system.terminate();
        });
    }
}

File: scala-demo/src/main/scala/demo2/GraphEx.scala
package demo2

import akka.actor.ActorSystem
import akka.stream.scaladsl._
import akka.stream.{ActorMaterializer, _}

import scala.concurrent.Await
import scala.concurrent.duration.Duration

// Important for using the FLowGraph DSL!
import akka.stream.scaladsl.GraphDSL.Implicits._

object GraphEx extends App {
  implicit val system = ActorSystem("demo2")
  implicit val materializer = ActorMaterializer()

  // create a stream for the following graph
  //                      f2
  // in -> f1 -> bcast /     \ merge -> f3 -> out
  //                   \  f4 /
  val sink = Sink.foreach(println)
  val g = RunnableGraph.fromGraph(GraphDSL.create(sink) { implicit b =>
    sink =>
      val in = Source(List(1, 2, 3))
      val bcast = b.add(Broadcast[Int](2))
      val merge = b.add(Merge[Int](2))

      val f1, f2, f3, f4 = Flow[Int].map(_ + 10)

      in ~> f1 ~> bcast ~> f2 ~> merge ~> f3 ~> sink
      bcast ~> f4 ~> merge
      ClosedShape
  })

  val res = g.run()
  Await.result(res, Duration.Inf)
  system.terminate
}
